src/ec2/model.ts

```diff
@@ -86,16 +94,19 @@ export class Ec2ConnectionManager {
         throw new ToolkitError(generalErrorMessage + message, errorInfo)
     }
 
-    protected async throwPolicyError(selection: Ec2Selection) {
-        const role = await this.ec2Client.getAttachedIamRole(selection.instanceId)
+    private async getPolicyErrorText(selection: Ec2Selection): Promise<string> {
+        try {
+            const role = await this.getAttachedIamRole(selection.instanceId)
```

Comment:
Making another (redundant) service call in logic that is trying to _build_ an error message, should be avoided. It's another place that can fail, while already handling a failure. It's also inefficient.

Can't `throwPolicyError` get enough info to make this decision without another service call?

Reply:
Yea, that makes a lot of sense. I think this is a symptom of the structure of the error handling. I restructured it so that not having a role is its own error. I think the new structure is a little cleaner. I had to refactor the tests to make it work with new structure.